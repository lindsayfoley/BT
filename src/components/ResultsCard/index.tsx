import { FunctionComponent } from 'react';
interface ISource {
  id: string;
  name: string;
}
export interface IArticle {
  title: string;
  author: string;
  description: string;
  publishedAt: string;
  url: string;
  source: ISource;
}

export const CHARACTER_LIMIT = 150;
export const truncateText = (text: string) => `${text.slice(0, CHARACTER_LIMIT)}...`;

const ResultsCard: FunctionComponent<IArticle> = (article: IArticle) => {
  const {
    title,
    author,
    description,
    publishedAt,
    url,
    source: { name, id },
  } = article;

  const truncatedDescription = description && truncateText(description);

  return (
    <>
      <article key={publishedAt}>
        <h3>{title}</h3>
        <span className="author">{author || name || id}</span>
        <p>
          <span className="description" dangerouslySetInnerHTML={{ __html: truncatedDescription }} />
          <a target="_blank" rel="noreferrer" href={url}>
            read more &gt;
          </a>
        </p>
      </article>
      <hr />
    </>
  );
};

export default ResultsCard;
