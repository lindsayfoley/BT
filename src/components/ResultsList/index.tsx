import { FunctionComponent } from 'react';
import ResultsCard, { IArticle } from '../ResultsCard';

export interface IResultsListProps {
  list: IArticle[];
}

const ResultsList: FunctionComponent<IResultsListProps> = ({ list }) => {
  if (!list || list.length === 0) {
    return null;
  }

  return (
    <section>
      <h2>{`Showing ${list.length} results:`}</h2>
      {list.map((article, index) => (
        <ResultsCard {...article} key={`${article.title}-${index}`} />
      ))}
    </section>
  );
};

export default ResultsList;
