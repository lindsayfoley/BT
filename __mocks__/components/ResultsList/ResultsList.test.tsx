import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import ResultsList from '../../../src/components/ResultsList';
import ResultsCard, { IArticle } from '../../../src/components/ResultsCard';

export const mockArticleList: IArticle[] = [
  {
    title: 'An earnest review of a robotic cat pillow',
    author: 'Brian Heater',
    description:
      'You don’t need Qoobo in your life. Nobody needs Qoobo, exactly. In fact, first reactions tend to range from befuddlement to bemusement. The robotic cat pillow doesn’t make a ton of sense on the face of it — in part because Qoobo has no face.',
    publishedAt: '2021-02-17T09:30:00-07:00',
    url:
      'https://techcrunch.com/2020/12/30/an-earnest-review-of-a-robotic-cat-pillow/',
    source: {
      id: '123456',
      name: 'Techcrunch',
    },
  },
  {
    title: 'Australia: Man held after armed raid in search of cat',
    author: 'BBC News',
    description: 'Tony Wittman allegedly stormed the shelter on Monday in full camouflage gear and pointed an assault rifle at a female worker.',
    publishedAt: '2021-01-17T09:30:00-05:00',
    url: 'https://www.bbc.co.uk/news/world-australia-55667023',
    source: {
      id: '789101',
      name: 'BBC News',
    },
  },
];

const setup = () => shallow(<ResultsList list={mockArticleList} />);

describe('<ResultsList />', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = setup();
  });

  it('renders without crashing', () => {
    expect(wrapper.isEmptyRender()).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a header displaying the correct number of search results', () => {
    expect(wrapper.find('h2').text()).toContain(mockArticleList.length);
  });

  it('should render a list of ResultsCard', () => {
    expect(wrapper.find(ResultsCard)).toHaveLength(mockArticleList.length);
  });
});
