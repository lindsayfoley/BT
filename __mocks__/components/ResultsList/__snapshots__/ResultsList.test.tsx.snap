// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResultsList /> renders without crashing 1`] = `
<section>
  <h2>
    Showing 2 results:
  </h2>
  <ResultsCard
    author="Brian Heater"
    description="You don’t need Qoobo in your life. Nobody needs Qoobo, exactly. In fact, first reactions tend to range from befuddlement to bemusement. The robotic cat pillow doesn’t make a ton of sense on the face of it — in part because Qoobo has no face."
    key="2021-01-17T09:30:00-05:00"
    publishedAt="2021-01-17T09:30:00-05:00"
    source={
      Object {
        "id": "123456",
        "name": "Techcrunch",
      }
    }
    title="An earnest review of a robotic cat pillow"
    url="https://techcrunch.com/2020/12/30/an-earnest-review-of-a-robotic-cat-pillow/"
  />
  <ResultsCard
    author="BBC News"
    description="Tony Wittman allegedly stormed the shelter on Monday in full camouflage gear and pointed an assault rifle at a female worker."
    key="2021-01-17T09:30:00-05:00"
    publishedAt="2021-01-17T09:30:00-05:00"
    source={
      Object {
        "id": "789101",
        "name": "BBC News",
      }
    }
    title="Australia: Man held after armed raid in search of cat"
    url="https://www.bbc.co.uk/news/world-australia-55667023"
  />
</section>
`;
